Rewrite BCPD description of the colour data format (#662)
gbdev/pandocs: src/Palettes.md

## LCD Color Palettes (CGB only)

> 

Author: Introduction? CGB has colours, palettes.

The old bit summing up the CRAM size was essentially doing this before. (Not that that was a good solution)

Colors on the Game Boy Color are stored as RGB555, meaning a single color is composed of three 5-bit components, one for each of red, green, and blue.

> Each 15-bit color occupies the lower part of a 16-bit “word”[^bit15]:

Author: You don't need to put quotes around 'word', it's a well established term in computing and (remarkably) this usage matches its lettery ancestor quite well.

```suggestion
Each 15-bit color occupies the lower part of a 16-bit word[^bit15]:
```

{{#include imgs/src/rgb555.svg}}

> The color palettes are stored in two dedicated banks of palette RAM (or <abbr title="Color RAM">CRAM</abbr> for *color RAM*), 64 bytes each[^cram_size]: one for background/window palettes and the other for OBJ palettes.

Author: Palettes haven't really been introduced at this point, but this reads like they have been.

> The two bytes of each color are stored **little-endian**, meaning that the low byte comes first.

Author: explicit byte order -- avoid confusion about most/least significant bit

```suggestion
The two bytes of each color are stored in **little-endian** byte order, meaning that the low byte comes first.
```

For example, the two palettes shown in the previous diagram would be stored like this:

{{#include imgs/src/color_ram.svg}}

> Unlike VRAM, OAM, or wave RAM, CRAM cannot be accessed directly, but instead through pairs of registers: one “data” register provides a one-byte view into CRAM, while the corresponding “index” register controls which CRAM address the data register is bound to.

Author: split long sentence
index before access

```suggestion
Unlike VRAM, OAM, or wave RAM, CRAM is not exposed in the memory map and cannot be accessed directly.
Instead, each bank of CRAM is accessed through a pair of registers: one register is used to select a CRAM address, and the other provides read/write access to the byte at that address.
```

> Much like VRAM, CRAM is inaccessible when the PPU is reading from it, that is, during [Mode 3](<#PPU modes>): writes are ignored, and reads return $FF.

Author: With the prior point starting "Unlike VRAM, ..." this one starting "Much like VRAM" makes a slightly off balance contrast.
You could put a "However, ..." at the start, but that reinforces it as an intentional contrast, which doesn't really make sense.

Reviewer: ```suggestion
Much like VRAM, the CRAM data registers are inaccessible when the PPU is reading from it, that is, during [Mode 3](<#PPU modes>): writes are ignored, and reads return $FF.
```

How about this?

Reviewer 2: This implies that writing to BCPD does not increment the index if autoinc is enabled in BCPI. But does it? It should be stated explicitly, though that's an edge case.

Reviewer: That edge case is explicitly stated in the description of BGPD; I think it's fine to keep it only there, as “the universe will always make a better idiot”.

Author: That does make more sense. The issue I had was pretty vague, so I'll consider that a fix.

```suggestion
Much like VRAM, the CRAM data registers are inaccessible when the PPU is reading from CRAM, that is, during [Mode 3](<#PPU modes>): writes are ignored, and reads return $FF.
```

[^bit15]:

> The 16th bit, bit 15, is **ignored** by the hardware.

Author: It's real though, right? I mean it's stored somewhere. The PPU "ignores" it, I suppose:

```suggestion
The 16th bit, bit 15, has no effect on the colour and is not used for any other purpose by the system.
```

Reviewer: I ended up going with “ignored by the rendering process”, which is more specific and doesn't imply that it isn't stored in CRAM.

> In discussion, that bit is generally clear (for example, the canonical pure white is `7FFF` and not `FFFF`), but the hardware treats both identically: it's fine to fill color RAM with $FF bytes to set it to all-white!

Reviewer 3: ```suggestion
Conventionally, that bit is generally clear (for example, the canonical pure white is `7FFF` and not `FFFF`), but the hardware treats both identically: it's fine to fill color RAM with $FF bytes to set it to all-white.
```

'In discussion' sounds a bit confusing here, imo

Reviewer 3: also the ! is a bit off tone

Reviewer: I wanted to avoid implying that it would be convention outside of the programs themselves. Perhaps "in writing"?

Reviewer 2: Proposal (that is definitely not meant as a final idea, but maybe some words can be taken up):

> When referring to a certain color, the 16-bit value is usually written with the highest bit set to 0, e.g. white will be referred to as $7FFF, not $FFFF.

(Also note the dollars.)

Reviewer: ```suggestion
Colors are conventionally referred to with that bit set to 0 (for example, the canonical pure white is `7FFF` and not `FFFF`), but the hardware treats both identically: thus, it's fine to fill color RAM with $FF bytes to set it to all-white.
```

I think it's better not to use the dollars here, as they'd imply numbers slightly more than colours. But that's fairly nitpicky.

Reviewer 2: Can we get an "e.g." after the colon?

Reviewer: Is this better?

Reviewer 2: I'm undecided on that one.

I would like to add one thing about the dollar sign: The notation without it could become ambiguous if no A-F digits are present. I think we need some way to mark that the color values (I'm not calling them colors!) are hexadecimal. Since we don't have the power (or need) for a unique notation like CSS's `#`, $ will do.

Reviewer 3: "referred to with that bit set to 0" sounds even more quirky, IMO

what about "The canonical form of a color value has that bit set to 0." ?

Reviewer: Again, no, since that implies that this is also the form that should be stuck to on the hardware itself. I'm trying to keep it clear that this is only the case in literature. (I switched to “that bit set to 0” because “that bit clear” is easy to misread, I've found.)

Author: I'm not sure if the point here is truly about this "convention" or if this note is intended as a way to illustrate that bit 15 is not used by the system.

- If it's about convention, I'd recommend removing the statement entirely. This is because (I think that) Pan Docs can/should avoid commenting on, relying on, or defining convention whereever possible/reasonable.
- If it's an illustrative example, remove the reference/s to convention/canon. Strictly speaking, the reader already has the information, so rephrase this as a caution/tip.
  - Possibly focusing more on (not) making assumptions about bit 15, one way or the other, rather than just about `$FFFF` and `$7FFF` both being white.

<hr>

dollar prefix:
I'd prefer to see the dollar sign here. Maybe this isn't as common as I'd expect, but to me these prefixes are just a way to tell the parser (human or otherwise) that the value is rendered in a particular way. So `0x` or `$` just signifies "hexadecimal nybbles follow", not really anything about the meaning of the data.

A (two's complement) signed integer must have the same issue -- in other words, the signed integer `$FFFF` is not rendered as a number. You have to expend mental effort to interpret it.

<hr>

> I'm not calling them colors!

I agree. I began using the term 'colour word' in my initial changes. I chose to not use 'colour value', though, because of the existing meaning/s of 'value' in colour theory / visual art / HSV.

Reviewer: I ended up making it clear that bit 15 is ignored *for rendering*, which then makes Antonio's original proposal unambiguous&mdash;so I went with that.

[^cram_size]:
2 bytes/color × 4 colors/palette × 8 palettes = 64 bytes.

### FF68 — BGPI (CGB Mode only): Background palette index

{{#bits 8 >
  "BGPI"  7:"Auto-increment" 5-0:"Address";
}}

- **Auto-increment**: `0` = Disabled; `1` = Enabled
- **Address**: Specifies which byte of BG Palette Memory can be accessed through
  [`BGPD`](<#FF69 — BGPD (CGB Mode only): Background palette data>)

Unlike `BGPD`, this register can be freely accessed outside VBlank and HBlank.

### FF69 — BGPD (CGB Mode only): Background palette data

As each color is two bytes in size, you must read/write this register *twice* to access a whole color.

> This is made much easier through the use of the address auto-increment: `BGPI`'s “address” field is automatically incremented (possibly wrapping around) after each write to this register, even if the write fails due to CRAM being inaccessible.

Reviewer 3: ```suggestion
This is made much easier through the use of the address auto-increment: `BGPI`'s "address" field is automatically incremented (possibly wrapping around) after each write to this register, even if the write fails due to CRAM being inaccessible.
```

> This is made much easier through the use of the address auto-increment: `BGPI`'s "address" field is automatically incremented (possibly wrapping around) after each write to this register, even if the write fails due to CRAM being inaccessible.

Author: "possible wrapping" makes it sound unreliable

You could say "mod 16" or (maybe more likely to ensure comprehension in this context) "bitwise `$1F`

```suggestion
This is made much easier through the use of the address auto-increment: `BGPI`'s "address" field is automatically incremented (wrapping around to zero after index 15) after each write to this register, even if the write fails due to CRAM being inaccessible.
```

Reads, however, never trigger auto-increment.

:::tip NOTE

> All background colors are initialized as white by the boot ROM, however it is a good idea to initialize all colors yourself, e.g. if implementing a soft-reset mechanic.

Reviewer 3: ```suggestion
All background colors are initialized as white by the boot ROM, however it is a good idea to initialize all colors yourself, e.g. by implementing a soft-reset mechanic.
```

_e.g.  if implementing ..._ doesn't sound good

Reviewer 4: "by" changes the meaning of the sentence. Maybe "for" would work better?

Reviewer: That would also change the meaning of the sentence. The point is that it's a good idea to set all palettes yourself if there's any chance that the init code may run more than once, such as if the game has a soft-reset mechanism.

Reviewer 4: What do you think of "for a soft-reset mechanic" rather than "for/by/if implementing a soft-reset mechanic"?

Reviewer 2: I like Isso's original proposal.

Author: Yes, changing it to "by" would make it have a different, also undesirable/confusing meaning. But the "if implementing" thing *is* odd, and the whole point being made is a bit strange, actually:

When *wouldn't* it be a good idea to initialise ~~all~~ the colours yourself? If they are all white you can't see anything.
Regardless of the behaviour of the boot ROM, it's a "good idea" to initialise ~~the~~some palettes, because you want your palette/s to be the ones being used. At least I think so.

[Note: I don't actually know how true the boot ROM statement is, I assume someone checked.]

I know this behaviour is specific to the BG colours, but this "tip" (and the similar one for OB) could be moved up to the general explanation above -- it's really not specific to the register/s.

Reviewer: Simpler games that only use e.g. a single colour palette generally don't bother initialising more than that, and I don't want to imply that doing so is wrong.

I have moved the tip to the introduction.

Author: Oh, that makes sense, for sure. I kept saying "all the colours" in my previous comment, I suppose just because the text does the same. It doesn't need to say "all".

I'm not sure what the intent/point of the tip is -- outside of the description of the system startup behaviour. Possibly:
- The system only initialises the palettes when it first starts up.
- The colours will all be white by default, so the software must set up at least one palette for any background tile(maps) to be visible.
- At least during the development process, you might want to initialise palettes that you're not using to something discernable, if only to assist in spotting errors/bugs.
- If you're relying on the palettes being all-white, don't forget to reset that as part of a soft-reset.

Just the fact:
```suggestion
All background colors are initialized as white by the boot ROM.
```

The software's responsibility:
```suggestion
All background colors are initialized as white by the boot ROM, so the software must set up any palettes it uses in order to have background tilemaps appear as anything other than solid white.
```

Reviewer 2: "as anything other than solid white." misses the point.

Author: OK. Please explain it to me.

> I'm not sure what the intent/point of the tip is -- outside of the description of the system startup behaviour.

Reviewer: My intent was indeed point 4, but I suppose it doesn't belong in Pan Docs much. Stating the hardware fact only should be sufficient here; we may even move it outside of a tip, though I'll check out the render to make that judgement.

:::

### FF6A–FF6B — OBPI, OBPD (CGB Mode only): OBJ palette index, OBJ palette data

These registers function exactly like BGPI and BGPD respectively; the 64 bytes of OBJ palette memory are entirely separate from Background palette memory, but function the same.

> Note that while 4 colors are stored per OBJ palette, color #0 is never used, as it's always transparent. It's thus fine to write garbage values, or even leave color #0 uninitialized.

Reviewer 3: ```suggestion
Note that while 4 colors are stored per OBJ palette, color #0 is never used, as it's always transparent. It's thus fine to write garbage values, or even leave it uninitialized.
```

### FF69 — BCPD/BGPD (CGB Mode only): Background color palette data / Background palette data

This register provides read/write access to the CGB background palette memory.

> The byte accessed is specified in [BCPS/BGPI](<#FF68 — BCPS/BGPI (CGB Mode only): Background color palette specification / Background palette index>).

Reviewer: ```suggestion
This register provides read/write access to the byte located at the address within the CGB's background palette memory specified in [BCPS/BGPI](<#FF68 — BCPS/BGPI (CGB Mode only): Background color palette specification / Background palette index>).
```

> The color data format is RGB555, meaning a single color is composed of three 5 bit components, one for each of red, green, and blue.

Reviewer: ```suggestion

The color data format is RGB555, meaning a single color is composed of three 5-bit components, one for each of red, green, and blue.
```

> Each 15 bit color occupies the lower part of a 16 bit word, with the 16th bit unused:

Reviewer: ```suggestion
Each 15-bit color occupies the lower part of a 16-bit word, with the 16th bit unused:
```

{{#bits 16 >
  "Channel"  15:"N/A" 14-10:"Blue" 9-5:"Green" 4-0:"Red";
}}

> As the colors are two bytes in size, you must read/write this register *twice* to access a whole color.

Reviewer: ```suggestion
As each color is two bytes in size, you must read/write this register *twice* to access a whole color.
```

> The low byte (bits 0-7) is stored first (lower) in memory -- with *litle-endian* byte order.

Reviewer: ```suggestion
The low byte (bits 0-7) is stored first (lower) in memory (*little-endian* byte order).
```

As each color is two bytes in size, you must read/write this register *twice* to access a whole color.

> This is made much easier through the use of the address auto-increment: `BGPI`'s "address" field is automatically incremented (wrapping around to zero after index 15) after each write to this register, even if the write fails due to CRAM being inaccessible.

Reviewer 2: This is plain wrong. There is nothing related to index 15.
- 8 palettes with 4 colors each
- 32 colors
- 64 bytes

Reviewer: ```suggestion
This is made much easier through the use of the address auto-increment: `BGPI`'s "address" field is automatically incremented (wrapping around from 63 back to 0) after each write to this register, even if the write fails due to CRAM being inaccessible.
```
